src/hotspot/share/opto/gcm.cpp

```diff
@@ -562,6 +562,35 @@ bool PhaseCFG::unrelated_load_in_store_null_block(Node* store, Node* load) {
   return false;
 }
 
+static bool already_enqueued(const Node_List& worklist_mem, const Node_List& worklist_store, Node* mem, PhiNode* phi) {
+  // mem is one of the inputs of phi and at least one input of phi is
+  // not mem. It's however possible that phi has mem as input multiple
+  // times. If that happens, phi is recorded as a use of mem multiple
+  // times as well. When PhaseCFG::insert_anti_dependences() goes over
+  // uses of mem and enqueues them for processing, phi would then be
+  // enqueued for processing multiple times when it only needs to be
+  // processed once. The code below checks if phi as a use of mem was
+  // already enqueued to avoid redundant processing of phi.
+  uint j = worklist_mem.size();
+  // The pair worklist_mem/worklist_store is used as a queue of pairs
+  // (mem,store) where mem is pushed to worklist_mem and store is
+  // pushed to worklist_store and store is a use of mem. Anytime a mem
+  // is pushed/popped to/from worklist_mem, a store has to be
+  // pushed/popped to/from worklist_store.
+  // If there are any use of mem already enqueued, they were enqueued
+  // last (all use of mem are processed in one go).
+  for (; j > 0; j--) {
+    if (worklist_mem.at(j-1) != mem) {
+      // We're done with the uses of mem
+      return false;
```

Comment:
Should this be `continue;` instead to make sure we scan all memory nodes?

Reply:
Seems my old comments have disappeared, and with them Roland's answers.

Roland said duplicates are ok. But if you @vnkozlov are asking, then @rwestrel needs to make a comment about why we can early exit here ;)

Reply:
Ah, it the comment is still here:
https://github.com/openjdk/jdk/pull/19791#issuecomment-2180754800

Reply:
I think the last statement in comment is not correct.
You can push several `(mem, store)` with the same memory. Then you pop only one top pair and start processing it which leads to different memory in `(mem, store)` pairs pushed on worklist. After processing some of pairs you may see original memory node again.

Reply:
Thank you, @eme64.  The last statement have to be adjusted based on Roland's answer you pointed.

Reply:
We start by pushing:
```
(null, initial_mem)
```
then we pop it and push all uses of `initial_mem` before we pop anything:
```
(initial_mem, store_1)
(initial_mem, store_2)
(initial_mem, store_3)
...
(initial_mem, store_n-1)
(initial_mem, store_n)
```
Then we pop a pair:
```
(initial_mem, store_n)
```
And we push uses of `store_n` only if it is a `MergeMem`
```
(initial_mem, store1)
(initial_mem, store2)
(initial_mem, store3)
...
(initial_mem, storen-1)
(store_n, store_n_1)
...
(store_n, store_n_n)
```
And we keep going.
What you're saying is we can end up with:
```
(mem1, ...)
(mem2, ...)
(mem1, ...)
```
For that `mem1` and `mem2` have to be a `MergeMem`.  So There has to be a cycle from `mem1` to `mem1` through `mem2` all of which are `MergeMem`. That can't happen right? Or am I missing something?

Reply:
> And we push uses of store_n only if it is a MergeMem
You are right, I missed `op == Op_MergeMem ` check.

Yes, I thought about loop since we check for Phi node. But we have `worklist_visited` check for `MergeMem` which will not allow to push the same `mem1` on work list. It look like you are right - there should be only sequential elements in work list for the same memory.
I withdraw my comment.